File: problem46/Solution.py
class Solution:
    def calcEquation(self, equations, values, queries):
        # Create a weighted graph (dict of dicts)
        graph = {}
        for equation, value in zip(equations, values):
            if equation[0] not in graph:
                graph[equation[0]] = {}
            if equation[1] not in graph:
                graph[equation[1]] = {}
            graph[equation[0]][equation[1]] = value
            # inverse value for a1 / a0
            graph[equation[1]][equation[0]] = 1 / value
        
        visited, valStack = [], [] # list for visited nodes, stack for values

        def dfs(node, target):
            if node == target: return valStack.pop()
            visited.append(node)
            for key, val in graph[node].items():
                if key in visited: continue
                valStack.append(val)
                value = dfs(key, target)
                if value: return value * valStack.pop()
            valStack.pop()
            return 0

        ans = []

        for query in queries:
            visited.clear()
            valStack.append(1)
            if query[0] not in graph or query[1] not in graph:
                ans.append(-1)
            else:
                val = dfs(query[0], query[1])
                ans.append(-1 if val == 0 else val)
        
        return ans

File: problem46/Solution.java
class Solution {
    // Create a map (HashMap) of HashMaps to represent the weighted graph
    Map<String, Map<String, Double>> graph = new HashMap<>();

    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        for (int i = 0; i < equations.size(); i++) {
            List<String> equation = equations.get(i);
            String var1 = equation.get(0);
            String var2 = equation.get(1);
            double value = values[i];

            graph.putIfAbsent(var1, new HashMap<>());
            graph.putIfAbsent(var2, new HashMap<>());

            graph.get(var1).put(var2, value);
            graph.get(var2).put(var1, 1 / value); // Add inverse value for backtracking
        }

        double[] results = new double[queries.size()];
        for (int i = 0; i < queries.size(); i++) {
            List<String> query = queries.get(i);
            String start = query.get(0);
            String end = query.get(1);

            if (!graph.containsKey(start) || !graph.containsKey(end)) {
                results[i] = -1;
                continue;
            }

            double result = dfs(start, end, new HashSet<>(), 1.0);
            if (result == 0) results[i] = -1;
            else results[i] = result;
        }
        return results;
    }

    private double dfs(String current, String target, Set<String> visited, double accumulatedValue) {
        if (current.equals(target)) return accumulatedValue;

        visited.add(current);

        for (Map.Entry<String, Double> neighbor : graph.get(current).entrySet()) {
            String var = neighbor.getKey();
            double val = neighbor.getValue();
            
            if (!visited.contains(var)) {
                double result = dfs(var, target, visited, accumulatedValue*val);
                if (result != 0) return result;
            }
        }

        return 0.0; // No path found
    }
}
